mirror_node_gateway.go

```diff
@@ -0,0 +1,99 @@
+package hedera
+
+/*-
+ *
+ * Hedera Go SDK
+ *
+ * Copyright (C) 2020 - 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+// Function to obtain the token relationships of the specified account
+func TokenReleationshipMirrorNodeQuery(networkUrl string, id string) (map[string]interface{}, error) {
+	tokenRelationshipUrl := buildUrlParams(networkUrl, "accounts", id, "tokens")
+	return makeGetRequest(tokenRelationshipUrl)
+}
+
+// Function to obtain account info for given account ID. Return the pure JSON response as mapping
+func AccountInfoMirrorNodeQuery(networkUrl string, accountId string) (map[string]interface{}, error) {
+	accountInfoUrl := buildUrlParams(networkUrl, "accounts", accountId)
+	return makeGetRequest(accountInfoUrl)
+}
+
+// Function to obtain balance of tokens for given contract ID. Return the pure JSON response as mapping
+func ContractInfoMirrorNodeQuery(networkUrl string, contractId string) (map[string]interface{}, error) { // nolint
+	contractInfoUrl := buildUrlParams(networkUrl, "contracts", contractId)
+	return makeGetRequest(contractInfoUrl)
+}
+
+// Function to obtain balance of tokens for given account ID. Return the pure JSON response as mapping
+func AccountBalanceMirrorNodeQuery(networkUrl string, accountId string) (map[string]interface{}, error) {
+	info, err := AccountInfoMirrorNodeQuery(networkUrl, accountId)
+	// in case of empty info we won't be able to map to string interface
+	if len(info) == 0 {
+		return nil, nil
+	}
+	return info["balance"].(map[string]interface{}), err
+}
+
+// Function to deduce the current network from the client as the network is ambiguous during Mirror Node calls
+func FetchMirrorNodeUrlFromClient(client *Client) string {
+	const localNetwork = "127.0.0.1"
+	const apiVersion = "/api/v1"
+	if strings.HasPrefix(client.GetMirrorNetwork()[0], localNetwork) {
+		return "http://" + localNetwork + ":5551" + apiVersion
+	} else {
+		// prefix is mainnet, testnet or previewnet
+		return "https://" + client.GetMirrorNetwork()[0] + apiVersion
+	}
+}
+
+// Make a GET HTTP request to provided URL and map it's JSON response to a generic `interface` map and return it
+func makeGetRequest(networkUrl string) (response map[string]interface{}, e error) {
```

Comment:
I don't really like that we are returning generic `interface`. One suggestion - to modify your function to return the response body contents directly and let the caller handle the JSON decoding.

```go
// Make a GET HTTP request to provided URL and return its body and any potential error
func makeGetRequest(networkUrl string) (body []byte, err error) {
    // Make an HTTP request
    resp, err := http.Get(networkUrl)
    if err != nil {
        return nil, err
    }
    defer resp.Body.Close()

    // Read the response body
    body, err = ioutil.ReadAll(resp.Body)
    if err != nil {
        return nil, err
    }

    // Check for non-200 status codes
    if resp.StatusCode >= 400 {
        return nil, fmt.Errorf("HTTP request failed with status code: %d", resp.StatusCode)
    }

    return body, nil
}
```

Reply:
We could ditch the `TokenRelationshipFromJson` and other similar methods if we are using `json.Unmarshal`

Reply:
Could be done when we introduce "full" queries to the mirror node